### ISSUE TITLE
OpenWhisk running on CloudFoundry

### ISSUE THREAD

**Issue by pjdurai** [apache/openwhisk#2109]
Greetings,

Currently OpenWhisk requires an Ubuntu Virtual Machine with Docker to get installed and running.
Is there any interest/efforts in getting it running in CloudFoundry's docker support?

CloudFoundry seems to support Docker (atleast 'docker build' and 'docker run' from the CF cli).

Is that support sufficient to get OpenWhisk running on CloudFoundry?

Any Ideas/Opinions on that from experts?

Thanks
pj

**Comment by style95** [comment]
I have few experience on CloudFoundry, and didn`t try to run OW on it.
But I have experience on running OW with Swarm.
You may refer to it.

Since CloudFoundry is PaaS, I guess it might provide docker in a similar way with Swarm.

In Swarm, it provides a single docker endpoint which accepts docker run request.
It will deploy a container on one of(few of) its agent machines.
Scheduling is managed by swarm manager.
So user can only care about a single docker endpoint to run container on docker cluster.

But in OpenWhisk, it is deployed via ansible which requires ssh access to host machine.
And Openwhisk Invoker communicates with local docker engine.
It requires docker unix socket as well as volume mapping of docker RootDir.
To find out the RootDir of docker, it uses docker remote api.
(Docker remote api is used in other parts also)

To sum up,

1. SSH access on docker host
2. Remote docker api should be enabled.
3. Access on host(volume mapping of docker RootDir)

There could be more things to consider, but in my case(Swarm), it was working with those things.
I am not sure these are allowed in CloudFoundry.

I hope it helps you.

And for the next time, you may want to join [slack](http://slack.openwhisk.org/) or subscribe mailing list: dev@openwhisk.incubator.apache.org.

**Comment by poblin-orange** [comment]
Another path could be deploying openwhisk on Bosh (if your considering on premise cloudfoundry, you have bosh director).
There is already : a docker bosh release wich could be leveraged, with swarm support.
https://github.com/cloudfoundry-community/docker-boshrelease

**Comment by krook** [comment]
I agree the appropriate integration here isn't wouldn't be to run OpenWhisk on top of Cloud Foundry, but to use the [BOSH](https://bosh.io/) deployment tool (closely associated with Cloud Foundry, but also able to deploy other things, like Kubernetes via [Kubo](https://pivotal.io/kubo)) as an option to deploy and manage OpenWhisk.
